catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,266 @@ import (
 	if err != nil {
-		return nil, err
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
 	}
-	stat, err := file.Stat()
+	return nil
+}
+
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx, err := newIndex(metas)
 	if err != nil {
-		_ = file.Close()
-		return nil, err
+		return err
 	}
-	if !stat.Mode().IsRegular() {
-		_ = file.Close()
-		return nil, os.ErrNotExist
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
+}
+
+func (s *LocalDirV1) BaseURL(catalog string) string {
+	return s.RootURL.JoinPath(catalog).String()
+}
+
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	return mux
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
+		serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader, ok := idx.Get(catalogFile, schema, pkg, name)
+	if !ok {
+		httpError(w, fs.ErrNotExist)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), indexReader)
+}
+
+func (s *LocalDirV1) catalogData(catalog string) (*os.File, os.FileInfo, error) {
+	catalogFile, err := os.Open(catalogFilePath(s.catalogDir(catalog)))
+	if err != nil {
+		return nil, nil, err
+	}
+	catalogFileStat, err := catalogFile.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+	return catalogFile, catalogFileStat, nil
+}
+
+func httpError(w http.ResponseWriter, err error) {
+	var code int
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		code = http.StatusNotFound
+	case errors.Is(err, fs.ErrPermission):
+		code = http.StatusForbidden
+	case errors.Is(err, ErrInvalidParams):
+		code = http.StatusBadRequest
```

Comment:
I don't see any calls to `httpError`, where we pass `ErrInvalidParams` in. But I think this is definitely something we should keep. Just need to make sure we actually check for unknown parameters pass it in.